Add GroupManagerWidget

tsalemink opened the pull request:
This PR adds the `GroupManagerWidget` class.

This widget takes a list of Zinc groups as an input and provides a GUI allowing the user to redefine a specific group based on the nodes/elements contained in each of the other groups in the list.

As shown in the following image, the widget contains a table where each line allows the user to define an operation to be performed on the chosen group in relation to another group defined over the same model. A new line is added to the end of the table each time the user selects a group for the operation at the bottom of the table.

![GroupManagerWidget](https://github.com/cmlibs-python/cmlibs.widgets/assets/82302710/6f6624bc-cfeb-4829-b2b7-bd7ba39a8f7a)

This widget emits a `group_updated` signal when the group operations are applied - to allow the caller to make live updates to any group visualisation it may use. The widget is non-modal by default.

hsorby commented on src/cmlibs/widgets/groupmanagerwidget.py:
```diff
@@ -0,0 +1,142 @@
+    group_add_not_group_elements, group_remove_not_group_elements
+
+
+DUMMY_TEXT = "---"
```
Let's call this `NULL_PLACEHOLDER`

hsorby commented on src/cmlibs/widgets/groupmanagerwidget.py:
```diff
@@ -0,0 +1,142 @@
+}
+
+
+class GroupManagerWidget(QtWidgets.QDialog):
```
As a widget I don't think we want to derive this from a `QDialog`.

tsalemink commented:
I've just made both the changes you requested and have squashed the commits together.

hsorby commented on src/cmlibs/widgets/groupmanagerwidget.py:
```diff
@@ -0,0 +1,140 @@
+    """
+    group_updated = QtCore.Signal()
+
+    def __init__(self, current_group, group_list, parent=None):
```
After having a bit of a think about this I think we need to change this around.  The `__init__` method should be defined as `def __init__(self, parent=None, current_group=None, group_list=None):`.  This should allow us to use qt creator and inherited widgets.

We would also need to add API to set the current group and group list.

hsorby commented:
I can see further development of this widget where we have fixed and flexible modes of using this widget.  I have also been throwing around some alternative names for this widget.  I am not super keen on `Group Manager Widget`, I am thinking maybe either `Group Operations Widget` or ` Group Manipulations Widget`.  Not committed to an alternative at this point, open to other options.

rchristie commented:
GroupEditorWidget?

tsalemink commented:
@hsorby, @rchristie.

I have updated the widget's API with a new constructor and setter methods for the `current_group` and `group_list` attributes.
I have renamed the widget to `Group Editor Widget`.

hsorby commented:
I think we might also need a `reset` API, after you have set the current group and changed the group list what do you do? Would use this method from the `__init__` method, also.

tsalemink commented:
Yes, I should have done that.

I believe this is essentially what the `_setup_widget` method does: set's the text for the current group,  removes the current group from the group list, and set's the group list as the items for the group combo boxes. So I should just be able to call `_clear_table` and then `_setup_widget` from the setter methods. Unless you'd prefer these methods were renamed?

hsorby commented:
If that works, I was thinking with the two API calls required to setup the data it might require and independent `reset` API call.

tsalemink commented:
Yep, good idea. I've added a `reset` method to the API that calls `_clear_table` and `_setup_widget`. This is called from the setter methods but could also be called externally with a single call to `reset`.

I've squashed the changes into the _[Update GroupManagerWidget API](https://github.com/cmlibs-python/cmlibs.widgets/pull/62/commits/900c3ae1a3c24ac9597d95c78e597bac42c5eb16)_ commit.
